packages/transaction-controller/src/TransactionController.ts

```diff
@@ -1235,6 +1207,103 @@ export class TransactionController extends BaseController<
+
+    await updateGas({
+      ethQuery: this.ethQuery,
+      providerConfig: this.getNetworkState().providerConfig,
+      txMeta: transactionMeta,
+    });
+
+    await updateGasFees({
+      eip1559: isEIP1559Compatible,
+      ethQuery: this.ethQuery,
+      getGasFeeEstimates: this.getGasFeeEstimates.bind(this),
+      txMeta: transactionMeta,
+    });
+  }
+
+  private getCurrentChainTransactionsByStatus(status: TransactionStatus) {
+    const chainId = this.getChainId();
+    return this.state.transactions.filter(
+      (transaction) =>
+        transaction.status === status && transaction.chainId === chainId,
+    );
+  }
+
+  private onBootCleanup() {
+    this.createApprovalsForUnapprovedTransactions();
+    this.loadGasValuesForUnapprovedTransactions();
+    this.submitApprovedTransactions();
+  }
+
+  /**
+   * Create approvals for all unapproved transactions on current chain.
+   */
+  private createApprovalsForUnapprovedTransactions() {
+    const unapprovedTransactions = this.getCurrentChainTransactionsByStatus(
+      TransactionStatus.unapproved,
+    );
+
+    for (const transactionMeta of unapprovedTransactions) {
+      this.processApproval(transactionMeta, {
+        shouldShowRequest: false,
+      }).catch((error) => {
+        if (error?.code === errorCodes.provider.userRejectedRequest) {
+          return;
+        }
+        /* istanbul ignore next */
+        console.error('Error during persisted transaction approval', error);
+      });
+    }
+  }
+
+  /**
+   * Update the gas values of all unapproved transactions on current chain.
+   */
+  private async loadGasValuesForUnapprovedTransactions() {
+    const unapprovedTransactions = this.getCurrentChainTransactionsByStatus(
+      TransactionStatus.unapproved,
+    );
+
+    const results = await Promise.allSettled(
+      unapprovedTransactions.map(async (transactionMeta) => {
+        await this.updateGasProperties(transactionMeta);
+        this.updateTransaction(
+          transactionMeta,
+          'TransactionController:loadGasValuesForUnapprovedTransactions - Gas values updated',
+        );
+      }),
+    );
+
+    results.forEach((result, index) => {
+      if (result.status === 'rejected') {
+        this.failTransaction(unapprovedTransactions[index], result.reason);
+        /* istanbul ignore next */
+        console.error(
+          'Error while loading gas values for persisted transaction',
+          result.reason,
+        );
+      }
+    });
+  }
+
+  /**
+   * Force to submit approved transactions on current chain.
+   */
+  private submitApprovedTransactions() {
+    const approvedTransactions = this.getCurrentChainTransactionsByStatus(
+      TransactionStatus.approved,
+    );
+    for (const transactionMeta of approvedTransactions) {
+      this.approveTransaction(transactionMeta.id).catch((error) => {
+        /* istanbul ignore next */
+        console.error('Error while submitting persisted transaction', error);
```

Comment:
Should we include the transaction IDs in these `console.error` calls for easier debug?

Reply:
Done in https://github.com/MetaMask/core/pull/1916/commits/9c26594d78fcda3e22e571a73b03d670a8469719